[scroll-animations] https://scroll-driven-animations.style/demos/3d-shoe-explorer/css/ animations break after navigating away then back to page

graouts opened the pull request:
#### c152324493769faa042332fd62826b37258d67bc
<pre>
[scroll-animations] <a href="https://scroll-driven-animations.style/demos/3d-shoe-explorer/css/">https://scroll-driven-animations.style/demos/3d-shoe-explorer/css/</a> animations break after navigating away then back to page
<a href="https://bugs.webkit.org/show_bug.cgi?id=284732">https://bugs.webkit.org/show_bug.cgi?id=284732</a>
<a href="https://rdar.apple.com/141528296">rdar://141528296</a>

Reviewed by Antti Koivisto.

The demo at <a href="https://scroll-driven-animations.style/demos/3d-shoe-explorer/css/">https://scroll-driven-animations.style/demos/3d-shoe-explorer/css/</a> shows a 3d model of sorts that is rendered via &lt;canvas&gt;.
To perform the rotation of the model, the demo uses a `requestAnimationFrame` loop to query the current time of scroll-driven CSS
animations in the document. The animation object itself is cached upon page load. As the page enters the back-forward cache, the
wrapper object itself is preserved, but the `CSSAnimation` is canceled during page teardown under `BackForwardCache::trySuspendPage()`.
As the page is restored from the page cache, the wrapper object will no longer provide a valid current time since its associated
animation has been canceled.

To address this we no longer cancel style-originated animations when entering the page cache.

Running the new test with multiple iterations uncovered an unsafe comparison in `StyleOriginatedAnimation::phaseWithoutEffect()`
between a `WebAnimationTime` and a `Seconds` value, so we also address this.

* LayoutTests/animations/no-cancelation-when-entering-page-cache-expected.txt: Added.
* LayoutTests/animations/no-cancelation-when-entering-page-cache.html: Added.
* Source/WebCore/animation/StyleOriginatedAnimation.cpp:
(WebCore::StyleOriginatedAnimation::phaseWithoutEffect const):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::cancelStyleOriginatedAnimations const):

Canonical link: <a href="https://commits.webkit.org/294899@main">https://commits.webkit.org/294899@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1823011757cb5797eb015846c1fc3fd3a9d37937

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103455 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23137 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13456 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108629 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54099 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105494 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23487 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31688 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78610 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106461 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18203 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93320 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58945 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18026 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11366 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53456 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/87821 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11426 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111008 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30602 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22567 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87607 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30963 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89519 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87248 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32123 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9834 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24942 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16776 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30530 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35842 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30330 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33661 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31891 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/Styleable.cpp:
```diff
@@ -315,6 +315,9 @@ OptionSet<AnimationImpact> Styleable::applyKeyframeEffects(RenderStyle& targetSt
 {
+    if (element.protectedDocument()->backForwardCacheState() != Document::NotInBackForwardCache)
+        return;
+
```
This could use a comment as running animations in bf cache sounds weird.

graouts replied:
Fixed this in the updated patch. Note that animations of all kinds are suspended while in the page cache, see `WebAnimation::suspend(ReasonForSuspension)` and `WebAnimation::resume()`.

graouts committed:
[scroll-animations] https://scroll-driven-animations.style/demos/3d-shoe-explorer/css/ animations break after navigating away then back to page
https://bugs.webkit.org/show_bug.cgi?id=284732
rdar://141528296

Reviewed by Antti Koivisto.

The demo at https://scroll-driven-animations.style/demos/3d-shoe-explorer/css/ shows a 3d model of sorts that is rendered via <canvas>.
To perform the rotation of the model, the demo uses a `requestAnimationFrame` loop to query the current time of scroll-driven CSS
animations in the document. The animation object itself is cached upon page load. As the page enters the back-forward cache, the
wrapper object itself is preserved, but the `CSSAnimation` is canceled during page teardown under `BackForwardCache::trySuspendPage()`.
As the page is restored from the page cache, the wrapper object will no longer provide a valid current time since its associated
animation has been canceled.

To address this we no longer cancel style-originated animations when entering the page cache.

Running the new test with multiple iterations uncovered an unsafe comparison in `StyleOriginatedAnimation::phaseWithoutEffect()`
between a `WebAnimationTime` and a `Seconds` value, so we also address this.

* LayoutTests/animations/no-cancelation-when-entering-page-cache-expected.txt: Added.
* LayoutTests/animations/no-cancelation-when-entering-page-cache.html: Added.
* Source/WebCore/animation/StyleOriginatedAnimation.cpp:
(WebCore::StyleOriginatedAnimation::phaseWithoutEffect const):
* Source/WebCore/style/Styleable.cpp:
(WebCore::Styleable::cancelStyleOriginatedAnimations const):

Canonical link: https://commits.webkit.org/294899@main
